File: python/src/contriwork_notifications/adapters/smtp.py
"""SMTP adapter — outbound email via aiosmtplib."""

from __future__ import annotations

from collections.abc import Sequence
from email.message import EmailMessage

import aiosmtplib

from ..adapter import AdapterDeliverResult, AdapterStatus
from ..errors import ErrorCode
from ..payload import Payload
from ..severity import Severity

_DEFAULT_TIMEOUT_S = 30.0

_GENERIC_FAILURE = "Notification delivery failed"
_GENERIC_AUTH = "Authentication rejected"


class SmtpAdapter:
    """SMTP delivery via aiosmtplib.

    Default transport is STARTTLS on port 587. For implicit TLS (port 465)
    pass ``use_tls=True`` and ``start_tls=False``. Both connect parameters
    and credentials are opaque to the package; the consumer loads them
    from its own config source.

    Email has no programmatic silent flag, so ``silent`` is a no-op
    (CONTRACT.md §Quiet hours flags this explicitly). The orchestrator
    still reports DELIVERED_SILENT for accounting; the wire payload is
    identical.
    """

    def __init__(
        self,
        host: str,
        port: int,
        from_addr: str,
        to_addrs: Sequence[str],
        *,
        username: str | None = None,
        password: str | None = None,
        use_tls: bool = False,
        start_tls: bool = True,
        name: str = "smtp",
        timeout_seconds: float = _DEFAULT_TIMEOUT_S,
    ) -> None:
        self._host = host
        self._port = port
        self._from_addr = from_addr
        self._to_addrs = list(to_addrs)
        self._username = username
        self._password = password
        self._use_tls = use_tls
        self._start_tls = start_tls
        self.name = name
        self._timeout = timeout_seconds

    async def is_available(self) -> bool:
        return bool(self._host and self._port and self._from_addr and self._to_addrs)

    async def deliver(
        self,
        severity: Severity,
        payload: Payload,
        silent: bool,
    ) -> AdapterDeliverResult:
        del severity, silent  # see class docstring; silent is a no-op
        msg = EmailMessage()
        msg["From"] = self._from_addr
        msg["To"] = ", ".join(self._to_addrs)
        msg["Subject"] = payload.title
        body = payload.body
        if payload.url is not None:
            label = payload.url_title or payload.url
            body += f"\n\n{label}: {payload.url}"
        msg.set_content(body)

        try:
            await aiosmtplib.send(
                msg,
                hostname=self._host,
                port=self._port,
                username=self._username,
                password=self._password,
                use_tls=self._use_tls,
                start_tls=self._start_tls,
                timeout=self._timeout,
            )
        except aiosmtplib.SMTPAuthenticationError:
            return AdapterDeliverResult(
                status=AdapterStatus.PERMANENT_FAILURE,
                error_code=ErrorCode.AUTH_FAILED,
                detail=_GENERIC_AUTH,
            )
        except (aiosmtplib.SMTPRecipientsRefused, aiosmtplib.SMTPSenderRefused):
            return AdapterDeliverResult(
                status=AdapterStatus.PERMANENT_FAILURE,
                error_code=ErrorCode.ADAPTER_INVALID_PAYLOAD,
                detail=_GENERIC_FAILURE,
            )
        except TimeoutError:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.TIMEOUT,
                detail=_GENERIC_FAILURE,
            )
        except (aiosmtplib.SMTPException, OSError):
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.UPSTREAM_ERROR,
                detail=_GENERIC_FAILURE,
            )

        return AdapterDeliverResult(status=AdapterStatus.DELIVERED)

File: typescript/src/adapters/smtp.ts
import nodemailer, { type SendMailOptions, type Transporter } from "nodemailer";

import type { Adapter, AdapterDeliverResult } from "../adapter";
import { AdapterStatus } from "../adapter";
import { ErrorCode } from "../errors";
import type { Payload } from "../payload";
import type { Severity } from "../severity";

const DEFAULT_TIMEOUT_MS = 30_000;
const GENERIC_FAILURE = "Notification delivery failed";
const GENERIC_AUTH = "Authentication rejected";

/**
 * Minimal subset of `nodemailer.Transporter` the adapter actually uses.
 * Tests inject a fake conforming to this shape; real production code uses
 * `nodemailer.createTransport(...)` which returns a wider Transporter.
 */
export interface SmtpTransporter {
  sendMail(options: SendMailOptions): Promise<unknown>;
}

export interface SmtpOptions {
  readonly host: string;
  readonly port: number;
  readonly fromAddr: string;
  readonly toAddrs: ReadonlyArray<string>;
  readonly username?: string;
  readonly password?: string;
  /** Implicit TLS (default false; use port 465 with this set to true). */
  readonly useTls?: boolean;
  /** STARTTLS upgrade on a plaintext connection (default true; port 587). */
  readonly startTls?: boolean;
  readonly name?: string;
  readonly timeoutMs?: number;
  /**
   * Optional custom transporter (for unit tests). Production code leaves
   * this undefined and the adapter creates its own via
   * `nodemailer.createTransport`.
   */
  readonly transporter?: SmtpTransporter;
}

interface NodemailerError {
  readonly code?: string;
  readonly message?: string;
}

/**
 * SMTP delivery via nodemailer. Default transport is STARTTLS on port 587;
 * pass `useTls: true` + `startTls: false` for implicit TLS (port 465).
 *
 * Email has no programmatic silent flag; `silent` is a documented no-op
 * and the wire payload is identical regardless. The orchestrator still
 * reports DELIVERED_SILENT for accounting.
 */
export class SmtpAdapter implements Adapter {
  readonly name: string;
  private readonly fromAddr: string;
  private readonly toAddrs: ReadonlyArray<string>;
  private readonly transporter: SmtpTransporter;
  private readonly timeoutMs: number;

  constructor(options: SmtpOptions) {
    this.name = options.name ?? "smtp";
    this.fromAddr = options.fromAddr;
    this.toAddrs = [...options.toAddrs];
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;

    if (options.transporter !== undefined) {
      this.transporter = options.transporter;
    } else {
      const transport: Transporter = nodemailer.createTransport({
        host: options.host,
        port: options.port,
        secure: options.useTls ?? false,
        requireTLS: options.startTls ?? true,
        connectionTimeout: this.timeoutMs,
        greetingTimeout: this.timeoutMs,
        socketTimeout: this.timeoutMs,
        ...(options.username !== undefined && options.password !== undefined
          ? { auth: { user: options.username, pass: options.password } }
          : {}),
      });
      this.transporter = transport;
    }
    // Validate required fields after capturing them so isAvailable can answer.
    this.requiredFieldsOk =
      Boolean(options.host) &&
      options.port > 0 &&
      Boolean(options.fromAddr) &&
      this.toAddrs.length > 0;
  }

  private readonly requiredFieldsOk: boolean;

  async isAvailable(): Promise<boolean> {
    return this.requiredFieldsOk;
  }

  async deliver(
    _severity: Severity,
    payload: Payload,
    _silent: boolean,
  ): Promise<AdapterDeliverResult> {
    let body = payload.body;
    if (payload.url !== undefined) {
      const label = payload.urlTitle ?? payload.url;
      body += `\n\n${label}: ${payload.url}`;
    }

    try {
      await this.transporter.sendMail({
        from: this.fromAddr,
        to: this.toAddrs.join(", "),
        subject: payload.title,
        text: body,
      });
    } catch (err) {
      const code = (err as NodemailerError | undefined)?.code;
      if (code === "EAUTH") {
        return {
          status: AdapterStatus.PermanentFailure,
          errorCode: ErrorCode.AuthFailed,
          detail: GENERIC_AUTH,
        };
      }
      if (code === "EENVELOPE") {
        return {
          status: AdapterStatus.PermanentFailure,
          errorCode: ErrorCode.AdapterInvalidPayload,
          detail: GENERIC_FAILURE,
        };
      }
      if (code === "ETIMEDOUT") {
        return {
          status: AdapterStatus.RetriableFailure,
          errorCode: ErrorCode.Timeout,
          detail: GENERIC_FAILURE,
        };
      }
      return {
        status: AdapterStatus.RetriableFailure,
        errorCode: ErrorCode.UpstreamError,
        detail: GENERIC_FAILURE,
      };
    }

    return { status: AdapterStatus.Delivered };
  }
}
